src/lib/Lower.hs

```diff
@@ -88,35 +87,26 @@ instance HoistableState LFS where
 instance GenericTraverser SimpIR UnitB LFS where
   traverseExpr expr = case expr of
     Op (TabCon ty els) -> traverseTabCon Nothing ty els
-    Hof (For dir ixDict (Lam body)) -> traverseFor Nothing dir ixDict body
+    Hof (For dir ixDict body) -> traverseFor Nothing dir ixDict body
     Case _ _ _ _ -> do
       Case e alts ty _ <- traverseExprDefault expr
       eff' <- foldMapM getEffects alts
       return $ Case e alts ty eff'
     _ -> traverseExprDefault expr
-  traverseAtom atom = case atom of
-    Lam _ -> do
-      traverseAtomDefault atom >>= \case
-        Lam (LamExpr (LamBinder b ty arr _) body@(Block ann _ _)) -> do
-          let eff' = case ann of BlockAnn _ e -> e; NoBlockAnn -> Pure
-          return $ Lam $ LamExpr (LamBinder b ty arr eff') body
-        _ -> error "Expected a lambda"
-    _ -> traverseAtomDefault atom
+  traverseAtom atom = traverseAtomDefault atom
 
 type Dest = Atom
 
 traverseFor
   :: Emits o => Maybe (Dest SimpIR o) -> ForAnn -> SAtom i -> LamExpr SimpIR i
   -> LowerM i o (SExpr o)
-traverseFor maybeDest dir ixDict lam@(LamExpr (LamBinder ib ty arr eff) body) = do
-  ansTy <- getTypeSubst $ Hof $ For dir ixDict $ Lam lam
+traverseFor maybeDest dir ixDict lam@(UnaryLamExpr (ib:>ty) body) = do
+  ansTy <- getTypeSubst $ Hof $ For dir ixDict $ lam
   ixDict' <- substM ixDict
   ty' <- substM ty
-  eff' <- substM $ ignoreHoistFailure $ hoist ib eff
-  let eff'' = extendEffRow (S.singleton InitEffect) eff'
```

Comment:
I don't know! I deleted it because we only used the effect in annotating the lambda, but now `For` takes a lambda-like term without any effect annotations. I don't understand the `Init` effect. AFAICT, `RunInit` is never used? @apaszke